### DISCUSSION TITLE
Explain difference between "installing" and "enabling" features.

### DISCUSSION THREAD

**Discussion by willnationsdev** [OrchardCMS/OrchardCore#8534] [Q&A] [answered]
The IFeatureEventHandler has separate install and enable handlers for before and after each event, but I'm not sure when I should choose one over the other. The Features admin only allows you to enable/disable features. I presumed that the first time a feature is enabled, it gets installed (so becomes part of the ShellDescriptor?) while the enable/disable state is only part of the ShellState. But then when I looked at ShellStateUpdater, it always updates the install/enable and uninstall/disable events at the same time rather than only conditionally triggering the install and uninstall events in special cases. So, what exactly is the difference?

**Comment by jtkech** [comment, marked as answer]
@willnationsdev not easy to read ;)

Yes, you're right, e.g the install state seems to change at the same time than the enabled state. Before we had different loading strategies for modules, e.g. we could install a module that would be dynamically compiled, so as i remember a module could be in an installed state, but not yet runtime compiled so not yet enabled. Just saw a `NonCompiledFeatureEntry` class that we normally don't need as we don't support dynamic compilation anymore, we would have to clean the code.

So, don't care about the install / uninstall states, just consider the enabled / disabled states and their related events.
